indexes/private.js

```diff
@@ -0,0 +1,148 @@
+  let encrypted = []
+  let canDecrypt = []
+
+  const debug = Debug('ssb:db2:private')
+
+  const encryptedFile = path.join(indexesPath(dir), 'encrypted.index')
+  const canDecryptFile = path.join(indexesPath(dir), 'canDecrypt.index')
+
+  function save(filename, seq, arr, cb) {
+    const buf = toBuffer(fic.compress(arr))
+    const b = Buffer.alloc(4 + buf.length)
+    b.writeInt32LE(seq, 0)
+    buf.copy(b, 4)
+
+    writeFile(filename, b, { fsyncWait: false })
+      .then(() => cb())
+      .catch((err) => cb(err))
+  }
+
+  function load(filename, cb) {
+    readFile(filename)
+      .then((buf) => {
+        const seq = buf.readInt32LE(0)
+        const body = buf.slice(4)
+
+        cb(null, { seq, arr: fic.uncompress(body) })
+      })
+      .catch((err) => cb(err))
+  }
+
+  function loadIndexes(cb) {
+    load(encryptedFile, (err, data) => {
+      if (err) return cb(err)
+
+      const { seq, arr } = data
+      latestSeq = seq
+      encrypted = arr
+
+      debug('encrypted loaded', encrypted.length)
+      debug('latest seq', latestSeq)
+
+      load(canDecryptFile, (err, data) => {
+        canDecrypt = data.arr
+
+        debug('canDecrypt loaded', canDecrypt.length)
+
+        cb()
+      })
+    })
+  }
+
+  loadIndexes(() => {})
+
+  function saveIndexes(cb) {
+    save(encryptedFile, latestSeq, encrypted, () => {
+      save(canDecryptFile, latestSeq, canDecrypt, cb)
+    })
+  }
+
+  const bKey = Buffer.from('key')
+  const bValue = Buffer.from('value')
+  const bContent = Buffer.from('content')
+  const StringType = 0
+
+  function reconstructMessage(data, unboxedContent) {
+    let msg = bipf.decode(data.value, 0)
+    const originalContent = msg.value.content
+    msg.value.content = unboxedContent
+    msg.meta = {
+      private: true,
+      originalContent,
+    }
+
+    const len = bipf.encodingLength(msg)
+    const buf = Buffer.alloc(len)
+    bipf.encode(msg, buf, 0)
+
+    return { seq: data.seq, value: buf }
+  }
+
+  function decrypt(data, streaming) {
+    if (bsb.eq(canDecrypt, data.seq) !== -1) {
+      let p = 0 // note you pass in p!
+
+      p = bipf.seekKey(data.value, p, bValue)
+      if (~p) {
+        const pContent = bipf.seekKey(data.value, p, bContent)
+        if (~pContent) {
+          const content = ssbKeys.unbox(bipf.decode(data.value, pContent), keys)
+          if (content) return reconstructMessage(data, content)
+        }
+      }
+    } else if (data.seq > latestSeq) {
+      if (streaming) latestSeq = data.seq
+
+      let p = 0 // note you pass in p!
+
+      p = bipf.seekKey(data.value, p, bValue)
+      if (~p) {
+        const pContent = bipf.seekKey(data.value, p, bContent)
+        if (~pContent) {
+          const type = bipf.getEncodedType(data.value, pContent)
+          if (type === StringType) {
+            encrypted.push(data.seq)
+
+            const content = ssbKeys.unbox(
+              bipf.decode(data.value, pContent),
+              keys
+            )
```

Comment:
Just checking - I don't want private box 1 baked into the implementation. Can you please make it an example implementation if an unboxer which is added and can later be extracted?

Also i think you should jump to making sure jump to making sure you can implement an. Unboxer for private groups. I found it a little tricky to implement and given private box was simple relatively. I think you would end up with a tidier setup if you made sure the complex case works

Reply:
This is a good suggestion. I created this PR as the first step to solving #3. It doesn't go all the way we need to add the key store and the boxers concept. The main point of this was to get the pipeline in, tets and then we can add the other stuff.